clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -79,39 +81,28 @@ export class HttpConnection implements IConnection {
                 // No need to add a connection ID in this case
                 this.url = this.baseUrl;
                 this.transport = this.constructTransport(TransportType.WebSockets);
+                // We should just call connect directly in this case.
+                // No fallback or negotiate in this case.
+                await this.transport.connect(this.url, transferFormat, this);
             } else {
-                let headers;
                 const token = this.options.accessTokenFactory();
                 if (token) {
-                    headers = {
+                    this.headers = {
                         ["Authorization"]: `Bearer ${token}`,
                     };
                 }
 
-                const negotiatePayload = await this.httpClient.post(this.resolveNegotiateUrl(this.baseUrl), {
-                    content: "",
-                    headers,
-                });
-
-                const negotiateResponse: INegotiateResponse = JSON.parse(negotiatePayload.content as string);
-                this.connectionId = negotiateResponse.connectionId;
-
+                await this.setNegotiationResponse(this.headers);
                 // the user tries to stop the the connection when it is being started
                 if (this.connectionState === ConnectionState.Disconnected) {
                     return;
                 }
-
-                if (this.connectionId) {
-                    this.url = this.baseUrl + (this.baseUrl.indexOf("?") === -1 ? "?" : "&") + `id=${this.connectionId}`;
-                    this.transport = this.createTransport(this.options.transport, negotiateResponse.availableTransports, transferFormat);
-                }
+                return this.createTransport(this.options.transport, transferFormat);
```

Comment:
You need to `await this.createTransport(...` instead of `return`ing it. When you have an `async` function in TypeScript, you don't return a `Promise<T>`, you return a `T` and the compiler wraps everything up into a `Promise` for you. Just like in C# (partly because C# was the inspiration for async/await in TypeScript, and later ECMAScript)

Reply:
If you do this, what you're actually going to end up returning is `Promise<Promise<void>>`. The problem is that JavaScript allows any object to be returned by any function, even if it's labelled `void`, so it's not an error to do this, it's just weird behavior.